pkg/hubble/relay/pool/manager_test.go

```diff
@@ -0,0 +1,632 @@
+					return FakeClientConn{}, nil
+				},
+			},
+			want: []Peer{
+				{
+					hubblePeer.Peer{
+						Name: "unreachable",
+						Address: &net.TCPAddr{
+							IP:   net.ParseIP("192.0.1.1"),
+							Port: defaults.ServerPort,
+						},
+					},
+					nil,
+				},
+			},
+		}, {
+			name: "1 reachable peer",
+			pcBuilder: testutils.FakePeerClientBuilder{
+				OnTarget: func() string {
+					return "fake"
+				},
+				OnClient: func() (peerTypes.Client, error) {
+					return &testutils.FakePeerClient{
+						OnNotify: func(_ context.Context, _ *peerpb.NotifyRequest, _ ...grpc.CallOption) (peerpb.Peer_NotifyClient, error) {
+							i := -1
+							cns := []*peerpb.ChangeNotification{
+								{
+									Name:    "reachable",
+									Address: "192.0.1.1",
+									Type:    peerpb.ChangeNotificationType_PEER_ADDED,
+								},
+							}
+							return &testutils.FakePeerNotifyClient{
+								OnRecv: func() (*peerpb.ChangeNotification, error) {
+									i++
+									switch {
+									case i >= len(cns):
+										return nil, nil
+									default:
+										return cns[i], nil
+									}
+								},
+							}, nil
+						},
+						OnClose: func() error {
+							return nil
+						},
+					}, nil
+				},
+			},
+			ccBuilder: FakeClientConnBuilder{
+				OnClientConn: func(target string) (ClientConn, error) {
+					var once sync.Once
+					return FakeClientConn{
+						OnGetState: func() connectivity.State {
+							once.Do(func() {
+								close(done)
+							})
+							return connectivity.Ready
+						},
+					}, nil
+				},
+			},
+			want: []Peer{
+				{
+					hubblePeer.Peer{
+						Name: "reachable",
+						Address: &net.TCPAddr{
+							IP:   net.ParseIP("192.0.1.1"),
+							Port: defaults.ServerPort,
+						},
+					},
+					FakeClientConn{},
+				},
+			},
+		}, {
+			name: "1 peer is deleted",
+			pcBuilder: testutils.FakePeerClientBuilder{
+				OnTarget: func() string {
+					return "fake"
+				},
+				OnClient: func() (peerTypes.Client, error) {
+					return &testutils.FakePeerClient{
+						OnNotify: func(_ context.Context, _ *peerpb.NotifyRequest, _ ...grpc.CallOption) (peerpb.Peer_NotifyClient, error) {
+							i := -1
+							cns := []*peerpb.ChangeNotification{
+								{
+									Name:    "reachable",
+									Address: "192.0.1.1",
+									Type:    peerpb.ChangeNotificationType_PEER_ADDED,
+								}, {
+									Name:    "reachable",
+									Address: "192.0.1.1",
+									Type:    peerpb.ChangeNotificationType_PEER_DELETED,
+								},
+							}
+							return &testutils.FakePeerNotifyClient{
+								OnRecv: func() (*peerpb.ChangeNotification, error) {
+									i++
+									switch {
+									case i >= len(cns):
+										return nil, nil
+									case i == len(cns)-1:
+										defer func() {
+											close(done)
+										}()
+										<-time.After(10 * time.Millisecond)
+										fallthrough
+									default:
+										return cns[i], nil
+									}
+								},
```

Comment:
nit (no change required): 

I think readability of the tests would benefit greatly if we would could extract the common logic of the different `FakePeerClient` implementations here. They are all similar, but also slightly different. Plus due to the fact that we have nested `func()`s and non-local effects (access to the `done` variable which is defined and accessed hundreds of lines of code away) makes code is also not easily readable.

Now I'm actually not sure if there is common logic that can really be extracted. So again: No change required. Better to have hard-to-read unit tests than no unit tests at all. This code provides a ton of value already!
But if there was a helper to implement custom `FakePeerClient` to be extracted without much effort here, that would make future maintenance of these unit tests much easier.

Reply:
I agree these tests are not the easier to read and it's mostly due to having three layers of mocked interfaces. I'm not really sure how could this be addressed besides adding helpers for the most common mock configs.